src/physics/em/KNInteractor.hh

```diff
@@ -0,0 +1,78 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file KNInteractor.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Macros.hh"
+#include "base/Types.hh"
+#include "physics/base/Interaction.hh"
+#include "physics/base/ParticleTrackView.hh"
+#include "physics/base/SecondaryAllocatorView.hh"
+#include "physics/base/Secondary.hh"
+#include "physics/base/Units.hh"
+#include "KNInteractorPointers.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Perform Compton scattering, neglecting atomic binding energy.
+ *
+ * This is a model for the discrete Compton inelastic scattering process. Given
+ * an incident gamma, it adds a single secondary (electron) to the secondary
+ * stack and returns an interaction for the change to the incident gamma
+ * direction and energy. No cutoffs are performed for the incident energy or
+ * the exiting gamma energy.
+ *
+ * The rejection rate increases with particles of lower energy. With a "max
+ * rejection loops" of 8 for an incident gamma at 10 keV (the lower cutoff),
+ * the rejection fraction is about 1 in 10000.
+ *
+ * \note This performs the same sampling routine as in Geant4's
+ *  G4KleinNishinaCompton, as documented in section 6.4.2 of the Geant4 Physics
+ *  Reference (release 10.6).
+ */
+class KNInteractor
+{
+  public:
+    // Construct sampler from shared and state data
+    inline CELER_FUNCTION KNInteractor(const KNInteractorPointers& shared,
+                                       const ParticleTrackView&    particle,
+                                       const Real3&            inc_direction,
+                                       SecondaryAllocatorView& allocate);
+
+    // Sample an interaction with the given RNG
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine& rng);
+
+    // >>> COMMON PROPERTIES
+
+    //! Minimum incident energy for this model to be valid
+    //! TODO: this isn't currently used.
+    static CELER_CONSTEXPR_FUNCTION real_type min_incident_energy()
+    {
+        return 10 * units::kilo_electron_volt;
+    }
+
+    // Maximum loops to allow over the rejection function
+    static CELER_CONSTEXPR_FUNCTION int max_rejection_loops() { return 256; }
```

Comment:
Not really, I went with an arbitrary base-2 number than base-10. I think we want something smaller than 1000 but I'm not sure that limit will ever get hit in practice. I did a very tiny bit of playing around with the rejection level as a function of energy; it seems like at the lower energy limit for the process (10 keV) the  only 1 in 10000 samples made it to 8 loops, and at higher energy the rejection rate decreases rapidly.

Honestly I think we should consider eliminating the rejection failure mechanism:
- Rejection failure means we get an "unphysical" result because we've selected the process and it's just the numerical method that broke.
- We should never get a nonpositive rejection probability for a sample, so it's just a matter of a few extra loops even in the worst case.
- In the worst case of one-in-a-billion tracks taking twice as many loops (or however long), it's no big deal because it's so infrequent.

I don't know if we need to characterize the rejection fraction per loop as a function of incident energy, or just say it's really unlikely and go from there.

@tmdelellis Any thoughts based on the recent DBRC rejection Shift cuda implementation?

Reply:
> I don't know if we need to characterize the rejection fraction per loop as a function of incident energy, or just say it's really unlikely and go from there.

Geant claims that EM physics for gammas works up to 100 TeV, although Klein-Nishina fitted XS data only within [10 keV, 100 GeV]. LHC data is always below 14 TeV, which is the limit reached by p-p collisions. If the rejection rate decreases with energy, then a test with 10 keV and a test with 1 TeV seems enough to document the decision and move on.

Reply:
The other thing is that it may be either to (analytically or experimentally) verify the acceptance fraction for this one KN sampling distribution, but that may not be the case for other processes.

I can say that I trust the physics and math and derivations done by the G4 team, but I have a lot less confidence in their implementation decisions (exemplified by the "is this hydrogen" check `z < 1.5`), so I'm loathe to accept undocumented constants or heuristics in their implementations.

Reply:
> but that may not be the case for other processes.

Oh yeah. I assumed that we would "trust but verify", and check every new process that we implement, since we would adapt the code a lot anyway. In that sense, trusting the physics and math is limited to the manual and code comments.

> so I'm loathe to accept undocumented constants or heuristics in their implementations.

Sure, otherwise we would incur into repeating any inefficiency or bugs Geant already has. The point is to try to make something better.

Reply:
Another data point.  The GeantV implementation seems to not have this iteration limit.  See KleinNishinaComptonModel::SampleReducedPhotonEnergyRej: 
https://gitlab.cern.ch/GeantV/geant/-/blob/master/physics/electromagnetic/models/src/KleinNishinaComptonModel.cc#L430

It also "seems" that this appears in Geant4 in the 3 Compton models.

Reply:
Thanks for the additional context, Philippe! I will delete the iteration limit.